ath79: add support for Meraki MR16

Hurricos opened the pull request:
Port device support for Meraki MR16 from the ar71xx target to ath79.

Specifications:

  * AR7161 CPU, 16 MiB Flash, 64 MiB RAM
  * One PoE-capable Gigabit Ethernet Port
  * AR9220 / AR9223 (2x2 11an / 11n) WLAN

Installation:

  * Requires TFTP server at 192.168.1.101, w/ initramfs & sysupgrade .bins
  * Open shell case and connect a USB to TTL cable to upper serial headers
  * Power on the router; connect to U-boot over 115200-baud connection
  * Interrupt U-boot process to boot Openwrt by running:
       tftpboot 0c00000 <filename-of-initramfs-kernel>.bin;
       bootm 0c00000;
  * Copy sysupgrade image to /tmp on MR16
  * sysupgrade -i /tmp/<filename-of-sysupgrade>.bin

Notes:

  - There are two separate ARTs in the partition (offset 0x1000/0x5000 and
    0x11000/0x15000) in the OEM device. I suspect this is an OEM artifact;
    possibly used to configure the radios for different regions,
    circumstances or RF frontends. We assume the first is the correct one
    as was the case with ar71xx.

  - kmod-owl-loader is still required to load the ART partition into the
    driver.

  - The manner of storing MAC addresses is identical to in ar71xx (in
    an mtd partition labeled 'mac').

  - Migrating from ar71xx has not been thorougly tested (sysupgrade
    requires --force), but partitioning is identical to 18.06 ar71xx, so it
    should not be a problem.

  - The MR12 should be able to be migrated in a nearly identical manner as
    it shares much of its hardware with the MR16.

  - Thank-you Chris B for copious help with this port.

Signed-off-by: Martin Kennedy <hurricos@gmail.com>

Hurricos commented:
To mention it: I squashed his commits because of the requirements in https://openwrt.org/submitting-patches, but @riptidewave93 contributed a significant amount of this patch. I brought over the partition table to get it booting, tried to get WiFi working and tinkered a bit with the duplicated ATH partition. He filled in all of the LEDs, standardized the device tree and made everything look nice. I also used his Aruba AP-105 as a blueprint for starting the dts file, as I came into this with no prior experience with writing .dts trees.

adrianschmutzler commented on target/linux/ath79/dts/ar7161_meraki_mr16.dts:
```diff
@@ -0,0 +1,178 @@
+
+			mac: partition@fd0000 {
+				label = "mac";
+				reg = <0xfd0000 0x10000>;
```
missing readonly

Hurricos replied:
This must be written by the user as it is not contained in a partition in OEM. As best I understand at this point, the OEM config pulls it from the ART. I could supply programs to modify the ART, but that creates yet another problem.

adrianschmutzler replied:
Well, if the OEM firmware puts it here it must get it from somewhere. I don't think reading from an undefined intermediate is desirable. We should read from the original source.

adrianschmutzler replied:
And if the user has to put it there, he can as well put it in uci config.

Hurricos replied:
This makes for sticky situations when attempting to put routers into recovery mode. I agree about reading from source if possible. I will dig later for the original commit which did this; I suspect the source MAC is in the OEM configuration of the board.

That said -- for what it's worth, this is how the Openwrt 18.06 and onward MR16 is actually configured, meaning that we now have to roll backwards to find which of two locations is the correct one if a user is upgrading from 18.06.

Hurricos replied:
> in the OEM configuration of the board

-- to be clear, in the middle of the firmware partition, in the OEM's root filesystem.

I will dump from an unflashed Meraki and attempt to correspond it.

adrianschmutzler replied:
> That said -- for what it's worth, this is how the Openwrt 18.06 and onward MR16 is actually configured, meaning that we now have to roll backwards to find which of two locations is the correct one if a user is upgrading from 18.06.

If we find the source, we read from the source, and everything is fine.

Hurricos replied:
As best I can tell, the OEM device has the hex in only one location:

```
$ binwalk -R "\x00\x18\x0a\x35\xa1" *
:
:
Scan Time:     2020-07-24 02:59:41
Target File:   /home/labby/mr16_lo_old/tmp/mtd2ro
MD5 Checksum:  c279a8cf2df7af12962c0c99444f8ad2
Signatures:    1

DECIMAL       HEXADECIMAL     DESCRIPTION
--------------------------------------------------------------------------------
102           0x66            Raw signature (\x00\x18\x0a\x35\xa1\x6c)
:
```
It's precisely at the beginning of the firmware so I like the idea of switching the location from OpenWrt 19 to OpenWrt 20. There is no separate configuration for the WiFi MACs.

Do you know how I might add code to migrate the MAC partition over to hex 0x66 of the original firmware so that users do not get locked out by misconfigured MACs after upgrading?

Hurricos replied:
In  8d699c6, I have moved the mac partition to 0x80000+64KB, and relinked the WiFis to pull their MAC from it as well.

adrianschmutzler replied:
> Do you know how I might add code to migrate the MAC partition over to hex 0x66 of the original firmware so that users do not get locked out by misconfigured MACs after upgrading?

With the partition read-only, no way. And I'd actually like it to remain read-only. One could do stuff like keeping the old "mac" partition and add a small piece of code to the lower 02_network section that just sets the MAC address (via uci) if something is found there. But I don't really like that; personally, I feel like "it should be done the proper way now", but that will of course be a show stopper for all ar71xx users.

adrianschmutzler replied:
Hmm, just looked here:
https://openwrt.org/toh/meraki/mr16#flash_layout

This clearly says that the device has a "board config" partition for 0x80000 to 0xa0000. Initial support completely ignored that; so, support for this device in ar71xx appears to be actually "broken".
Despite, certain areas are not partitioned at all.

I'd thus at least add this partition from 0x80000 to 0xa0000, and call it "config". I also do not see any reason for that strange "rootfs first" partitioning there.

Hurricos replied:
> But I don't really like that; personally, I feel like "it should be done the proper way now", but that will of course be a show stopper for all ar71xx users.

A thought:

If someone is already running ar71xx, `0xbf08000 + 0x20000` is already unlocked, because it's `rootfs`. The process for writing the MAC address would be to, during stage2 (with rootfs unmounted), do:

`mtd erase $ROOTFS_MTD; echo -ne '<\x eth mac address>' | dd of=$ROOTFS_MTD bs=1 seek=$((0x66)) count=12`

If it's possible to script this into the sysupgrade between the two versions only, then I think it would be ideal.  I just don't see an obvious way to do that.

adrianschmutzler replied:
I'm not sure yet whether I like this. In my perception, the ar71xx support still feels "just broken", and I hesitate to add further hacks now to clean this up.
For the solution you propose, the problem is that the sysupgrade code run is that from the old version, isn't it? So, one would have to upgrade into an intermediate version before finally going into clean ath79.
In addition, if we change to denx,uimage as proposed, sysupgrade won't work anyway; so, maybe we just don't have to solve this problem, and ath79 needs to be flashed via TFTP again and fine.

adrianschmutzler replied:
BTW, if you aim at blocking the "old" mac partition anyway, the easiest way would be to just flash ath79 and then install kmod-mtd-rw via opkg.

This would allow users with old config to remove the read-only on the config partition, and write their address to it, while theoretical "new" users with intact config partition would not be affected by the problems from the past.

Hurricos replied:
> kmod-mtd-rw

This is good. I'm also considering instances where MR16s are not physically and changing the MAC address will change the IP address. Since mtd-rw gives us userspace writes to mtd, it'd be enough to add to the configuration during sysupgrade; I can provide a script to drop into /lib/upgrade, which provides a safe way to carry the present MAC addresses into uci, and assert that the configuration will be saved (or abort and reboot at any failures); I'd post the script in the wiki article.

Hurricos replied:
For posterity:

I did sleuthing to find what happened with the original MAC address setup. Chris B realized that u-boot was pulling from `0xbfff0000`, the beginning of the ART partition, even if he did not realize Meraki was storing it in the first 64K of their own partitions (which we overwrite). Here's what `u-boot` lets us do with progmac (which does not actually write anything back to flash, probably because it had not previously been `erase`d):

```
ar7100> progmac
Usage:
progmac <serno> - Program the MAC addresses
ar7100> progmac 00:11:22:33:44:55

First 0xff last 0xff sector size 0x10000                                                                                          255
write addr: bfff0000
```

This is the code from `build-dir/target/linux/ar71xx/files/arch/mips/ath79/mach-mr16.c` from [one of Chris B's MR16 LEDE releases](https://github.com/riptidewave93/Openwrt-MR12-MR16/archive/1.1Beta.tar.gz), showing that he uses what U-boot uses:

```
static void __init MR16_setup(void)                                             
{                                                                               
        u8 *mac = (u8 *) KSEG1ADDR(0xbfff0000);                                 
                                                                                
        /* Bring up MDIO */                                                     
        ath79_register_mdio(0,0x0);                                             
                                                                                
        /* 1GB POE Port */                                                      
        ath79_init_mac(ath79_eth0_data.mac_addr, mac, 0);           
```

adrianschmutzler commented on target/linux/ath79/generic/base-files/etc/board.d/01_leds:
```diff
@@ -168,6 +168,13 @@ glinet,gl-mifi)
+	ucidef_set_led_netdev "wan" "WAN" "$boardname:green:wan" "eth0"
+	ucidef_set_led_wlan "wlan1" "WLAN1" "$boardname:green:wifi1" "phy0assoc"
+	ucidef_set_led_wlan "wlan2" "WLAN2" "$boardname:green:wifi2" "phy0assoc"
+	ucidef_set_led_wlan "wlan3" "WLAN3" "$boardname:green:wifi3" "phy0assoc"
```
What's that supposed to do?

Hurricos replied:
https://github.com/btvmeshnet/openwrt/pull/1#pullrequestreview-451141038

adrianschmutzler replied:
I can see that.

My question was referring to the strange situation that you have two WiFi interfaces, four wifi LEDs, and set up all of them to only one phy but three with a different trigger than the fourth one.
Despite, what's phy0assoc doing, I've never seen this so far.

Hurricos replied:
Fair. it's intended to be an RSSI bar in OEM, but I'd have used phy0rx / phy0tx / phy1rx / phy1tx for each of WLAN LEDs 1-4. 

I will have to commit and re-squash to fix this, yes? I'd be putting the new squash of commits into a different pull request, then?

adrianschmutzler replied:
> I will have to commit and re-squash to fix this, yes? I'd be putting the new squash of commits into a different pull request, then?

Squash your changes, and force-push them here again. Don't open a new PR, we will just lose the discussion and have 10 PRs for the same subject after all.

> Fair. it's intended to be an RSSI bar in OEM, but I'd have used phy0rx / phy0tx / phy1rx / phy1tx for each of WLAN LEDs 1-4.

Just looked at the manual, yes, it's used as an RSSI bar, but we have two interfaces, so we cannot really implement this properly. Therefore, I typically prefer to have this unset, so the user can define it to his/her need.

Hurricos replied:
> Squash your changes, and force-push them here again. 

OK, great. I'll have those after work.

Hurricos replied:
I believe I have correctly reconfigured these to be empty. This awaits your review.

adrianschmutzler commented on target/linux/ath79/generic/base-files/etc/hotplug.d/firmware/10-ath9k-eeprom:
```diff
@@ -131,6 +131,7 @@ case "$FIRMWARE" in
 "ath9k-eeprom-pci-0000:00:11.0.bin")
 	case $board in
 	buffalo,wzr-hp-ag300h|\
+	meraki,mr16|\
```
ar71xx is actually reading from 0xff1000 here, you are reading from 0xfe1000. The equivalent is true for the other interface.
Despite, you don't set up a MAC address here manually, in contrast to ar71xx.

So, do you have correct MAC addresses on the WiFi interfaces right now?

Hurricos replied:
> ar71xx is actually reading from 0xff1000 here, you are reading from 0xfe1000. The equivalent is true for the other interface.

I mention this in the commit message, but was incorrect about which one ar71xx uses by default.

Meraki supplies four ARTs for each of two PHYs, all of which have correct checksums. Both sets of ARTs are functional but have slightly different properties (target powers in particular are higher in the upper 64KB). I was suspecting this was a change for different regions, as there are A and B models of the Meraki MR16 with otherwise identical hardware, but neither ART suggests a region.

Hurricos replied:
>if you remove the setup in DTS.

I'm sorry, I don't follow. The DTS must indicate where the ART partition is to protect it anyways. We could have this point directly to its location in firmware as you mention, but other models with AR922X and no separate EEPROM chip do not do this.

adrianschmutzler replied:
> The value coming from the ART is available via /sys/kernel/debug/ieee80211/phy0/ath9k/base_eeprom, and is wrong and the same on all MR16's:

Well, I'd have just checked the flash at 0x1006, 0x5006, 0x11006 and 0x15006, but if that leads to the same result ...

> Meraki supplies four ARTs for each of two PHYs, all of which have correct checksums. Both sets of ARTs are functional but have slightly different properties (target powers in particular are higher in the upper 64KB). I was suspecting this was a change for different regions, as there are A and B models of the Meraki MR16 with otherwise identical hardware, but neither ART suggests a region.

Is there a GPL source or something where one could look what's actually used?

Hurricos replied:
Not quite. It's offset in the Modal EEPROM for each PHY as per the AR92XX document I would have attached if I could. It's since been removed from the Internet Archive.

I have no longer been able to find the MR16 source. It was released a while ago.

adrianschmutzler replied:
> Not quite. It's offset in the base EEPROM for each PHY as per the AR92XX document I would have attached if I could. It's since been removed from the Internet Archive.

Ah, was mixing stuff up with ath10k a little.

Hurricos replied:
Since the vendor provided misconfigured ARTs I'm not going to do anything about this (most end users do not know how to modify AR92xx ARTs or even read them).

adrianschmutzler commented on target/linux/ath79/dts/ar7161_meraki_mr16.dts:
```diff
@@ -0,0 +1,178 @@
+		led-boot = &power_green;
+		led-failsafe = &power_orange;
+		led-running = &power_green;
+		led-upgrade = &power_green;
```
Radio Power
Off: MR16 is off
Solid Orange: MR16 is booting or trying to fi nd a path to the internet
Solid Green: MR16 is fully operational and connected to the network
Flashing Orange: Firmware is upgrading
Flashing Green: Error state. May indicate bad gateway or other routing fault 

So, if flashing green indicates error, you shouldn't use that for boot, but use orange for boot and upgrade.

Hurricos replied:
Tweaked this. I believe I understood your meaning. I was not able to find the definitions required to cause flashing, however.

adrianschmutzler commented on target/linux/ath79/image/generic.mk:
```diff
@@ -999,6 +999,21 @@ define Device/librerouter_librerouter-v1
+  IMAGE/rootfs.bin := append-rootfs | pad-rootfs
+  IMAGE/sysupgrade.bin := append-rootfs | pad-rootfs | pad-to $$$$(ROOTFS_SIZE) | append-kernel | append-metadata | check-size
+  IMAGES := kernel.bin rootfs.bin sysupgrade.bin
+  SUPPORTED_DEVICES += meraki_mr16
```
Just "mr16", that's the string from ar71xx.

Hurricos replied:
I see. That may explain the `sysupgrade` refusal.

Hurricos replied:
Fixed in 8d699c6.

adrianschmutzler replied:
Depending on how we solve the MAC issue, we might drop the SUPPORTED_DEVICES eventually, so leave this open as a reminder.

Hurricos replied:
I should undo my fix in 8d699c6 so this check fails on upgrading from ar71xx.

`platform_check_image` should fail here, because right now the upgrade step described in the [most recent revision](https://github.com/openwrt/openwrt/commit/359b1ed3f1dac316f4223bd1a4368034951d1d38) demands the user's direct intervention to
- (step 5) set the kernel address in u-boot config, and
- (step 6) to modify `/lib/upgrade/comon.sh` to pad with 128K zeroes to fix alignment.

The latter we could solve with a second target image only for upgrading, with a 128K zero pad to the image, but the former we cannot fix without user intervention (or doing something very bad like extending the firmware partition into the u-boot config region).

Let me know what you think.

adschm replied:
Have a look at what the new DEVICE_COMPAT_VERSION makes possible here:
https://github.com/openwrt/openwrt/pull/3345/commits/06336990c55d356abb30d4f4dac1dab2c002eff7#diff-0835d68ee1baa61ee8893cec8a75f77eR49

Hurricos replied:
OK. Updated and built, and, as expected, my MR16 flashed with previous build rejects it due to compat_version being unset.

After sysupgrading -F it, it does like it, and `uci show system.@system[0].compat_version` is 2.0.

adrianschmutzler commented on target/linux/ath79/image/generic.mk:
```diff
@@ -999,6 +999,21 @@ define Device/librerouter_librerouter-v1
+  DEVICE_MODEL := MR16
+  ROOTFS_SIZE := 13440k
+  IMAGE_SIZE := 15680k
+  DEVICE_PACKAGES := kmod-spi-gpio kmod-owl-loader
```
I'm looking at kmod-spi-gpio for the first time, do we need it here? Most probably that's already enabled elsewhere, otherwise one of the other devices would require it as well ...

Hurricos replied:
Removed. Doesn't seem necessary, no.

adrianschmutzler commented on target/linux/ath79/image/generic.mk:
```diff
@@ -999,6 +999,21 @@ define Device/librerouter_librerouter-v1
+  IMAGE_SIZE := 15680k
+  DEVICE_PACKAGES := kmod-spi-gpio kmod-owl-loader
+  IMAGE/kernel.bin := append-kernel
+  IMAGE/rootfs.bin := append-rootfs | pad-rootfs
```
What are these two images good for? They are not mentioned in the flashing instructions.

Hurricos replied:
I need to rewrite the commit; sadly the flashing process listed there is inaccurate for an OEM MR16. The user must also set u-boot variables for where to boot from by default; in the past these images were used for initial installation, but that may no longer be necessary with a dts rather than a mach. I'm not sure. I'll test again from an OEM and rewrite the instructions.

https://github.com/btvmeshnet/openwrt/pull/1#discussion_r456921586 for reference.

adrianschmutzler replied:
I was additionally startled by the fact that you just define a single firmware partition now, but still do have a fixed offset for append-kernel in sysupgrade recipe. I wonder whether that breaks at some point, or will be irrelevant.

Hurricos replied:
I see this. I've unraveled the firmware partition in the .dts into a rootfs and kernel partition, and restructured the image builder to provide for a fixed offset for the kernel so that the instructions are applicable without conflicting with the dts.

adrianschmutzler replied:
As stated elsewhere already, I wonder whether this "rootfs first" approach is actually, or whether we could just use a single firmware partition with one of the splitters instead.

Hurricos replied:
I think the reason for the "rootfs first" approach is because the kernel offset has to be fixed for u-boot, but I'm not sure.

adrianschmutzler replied:
And that's why I was startled by you just moving that +64k for adding the mac partition at 0x80000.

Hurricos replied:
I've followed your lead here -- the kernel is first as it should be, and we preserve the Meraki config partition, and now have a tested (but not officially supported) solution for rearranging things as needed to migrate ar71xx to ath79 without opening the case, so I'd like to call this one done.

adrianschmutzler commented on target/linux/ath79/image/generic.mk:
```diff
@@ -999,6 +999,21 @@ define Device/librerouter_librerouter-v1
+  DEVICE_PACKAGES := kmod-spi-gpio kmod-owl-loader
+  IMAGE/kernel.bin := append-kernel
+  IMAGE/rootfs.bin := append-rootfs | pad-rootfs
+  IMAGE/sysupgrade.bin := append-rootfs | pad-rootfs | pad-to $$$$(ROOTFS_SIZE) | append-kernel | append-metadata | check-size
```
Please wrap so one line contains max. 80 characters. Hanging indent with one tab.

Hurricos replied:
Done in 8d699c6.

Hurricos commented:
Sorry for the extra version. Somehow the very much-needed comment about setting the bootcmd before booting into the initramfs for a sysupgrade was missing.

adrianschmutzler commented on target/linux/ath79/dts/ar7161_meraki_mr16.dts:
```diff
@@ -0,0 +1,189 @@
+
+	chosen {
+		bootargs = "console=ttyS0,115200";
+	};
```
This can be dropped, it's already in ar7100.dtsi.

Hurricos replied:
Removed.

adrianschmutzler commented on target/linux/ath79/dts/ar7161_meraki_mr16.dts:
```diff
@@ -0,0 +1,189 @@
+			gpios = <&gpio 4 GPIO_ACTIVE_LOW>;
+		};
+
+		power_orange: power_orange {
```
Actually, the DT _label_ should have a "led_" prefix, as it's used in a global context, so use:
`led_power_orange: power_orange {`

Same for green power led.

adrianschmutzler commented on target/linux/ath79/dts/ar7161_meraki_mr16.dts:
```diff
@@ -0,0 +1,189 @@
+					reg = <0x000000 0xd20000>;
+				};
+
+				partition@d20000 {
```
So, this will effectively move the kernel partition back by 64k compared to ar71xx. If that doesn't matter, i.e. the kernel partition's position is not required to be fixed, can't we just use some splitter here and go back to dynamic partitioning?

Hurricos replied:
Could you point me to some documentation or some other keyword I could use to peruse DTS files to see syntax?

Hurricos replied:
This is specifically in context of the word "splitter".

adrianschmutzler replied:
I haven't reviewed a case so far where kernel wasn't first in the firmware partition. If kernel is first, its start offset of course is fixed by the firmware partitions start, and then positioning the rootfs will just be matter of kernel size. This rootfs positioning can then be done by one the splitters here: https://github.com/openwrt/openwrt/tree/master/target/linux/generic/files/drivers/mtd/mtdsplit

Selection is done by adding the corresponding compatible to the firmware partition. This essentially is a mechanism to remove the necessity to resize kernel partitions manually from time to time when the kernel has grown beyond partition limits.

In your case, _I don't think_ this will make sense, as if the kernel partition is second, its offset cannot be moved around as for the rootfs in my earlier example; and I also don't know whether the splitters will work in that case (swapped order for kernel/rootfs).
For now, it might just work to implement fixed partitions with the correct offset.
(This will work until the kernel outgrows its partition size of about 2.2 MiB (I don't know how big the kernel is on ath79/generic ATM, but for ath79/nand several devices went beyond the 2 MiB limit). After that limit is reached, you will either have to implement a multi-stage bootloader that will be able to start the actual kernel from another position, or have the device set to DEFAULT := n and will have to build with local config manually again to reduce kernel size. Long story, but the essence is: If you are already close to that limit, it might be better to implement a decent solution right away.)

Hurricos replied:
All makes sense. I removed the forced kernel-last config order from the makefile  and used compatible='denx,uboot'. Works fine.

adrianschmutzler replied:
Interesting. So, where does the bootloader learn the kernel offset from then?

Hurricos replied:
We have to set it in u-boot manually; from the commit: `setenv bootcmd bootm 0xbf0a0000; saveenv;`

Hurricos replied:
This should be resolved.

adrianschmutzler commented on target/linux/ath79/image/generic.mk:
```diff
@@ -999,6 +999,20 @@ define Device/librerouter_librerouter-v1
+  DEVICE_PACKAGES := kmod-spi-gpio kmod-owl-loader
+  IMAGE/sysupgrade.bin := append-rootfs | pad-rootfs | \
+	pad-to $$$$(ROOTFS_SIZE) | append-kernel | append-metadata | check-size
+  IMAGES := sysupgrade.bin
```
IMAGES is set to the default and can be dropped.

Hurricos replied:
Done.

adrianschmutzler commented on target/linux/ath79/dts/ar7161_meraki_mr16.dts:
```diff
@@ -0,0 +1,181 @@
+				compatible = "denx,uimage";
+			};
+
+			legacy_mac: partition@a0000 {
```
@fd0000

adrianschmutzler replied:
And the "legacy_mac:" DT label can be dropped, as it's not used.

Hurricos replied:
Done.

adrianschmutzler commented:
That Wiki page is quite interesting. Is it possible that this device initially had some kind of dualboot/dual firmware configuration?

This whole flashing procedure there looks like it was done in a hustle and nobody cared about the original setup. I wonder whether it would be possible to boot with original bootloader instead just overwriting the address in the first place.

Hurricos commented:
My / our use case for the MR16 requires that we do not pull them down from where they are located, nor lose their configuration.

Ultimately we should move the MAC partition to make things easier for new users and possibly explore what other data lies in the OEM MR16 config partition.

I had success with `sysupgrade -F` from the OpenWrt v18.06.2 release for the MR16 to a build which has the beginning of the image remaining at `0xbf080000`. Configuration was kept. I believe the change in file type trips platform_check_image in platform.sh, which is why -F is required. However, sysupgrading with the `firmware` partition at `0xbf0a0000` doesn't work because the pre-sysupgrade system writes it to its location for the `firmware` partition, `0xbf08000`.

I couldn't finish this patch without having a serial-cable-less sysupgrade from v19 ar71xx to v20 ath79.

I've devised the following process:

- Installing `kmod-mtd-rw`
- Installing `uboot-envtools`
- Manually configuring and saving the u-boot env [using this guide](https://openwrt.org/docs/techref/bootloader/uboot.config), to point the `bootcmd` to 0xbf0a0000
- Manually modifying `/lib/upgrade/platform.sh` to make `platform_do_upgrade` pad the sysupgrade image with 128K zeroes (plus perhaps MAC address from the old MAC partition) as it feeds it to `mtd write`. I think this should be possible but need to review the sysupgrade code more closely to verify handling of data at the end of the `mtd write` process.
- Running `sysupgrade -c -F /tmp/<sysupgrade>.bin`.

I'll try to finish tonight the remaining requests of the patch. I need to verify that the above works with the final patch.

rotanid commented:
regarding the MAC of this device please also see the following PR from 2017 where the MAC was discussed without an outcome:
https://github.com/lede-project/source/pull/1408

xabolcs commented:
`DEVICE_COMPACT_VERSION` and `DEVICE_COMPACT_MESSAGE` is now supported by trunk, and it will very helpful here, IMHO.

adrianschmutzler commented:
> `DEVICE_COMPACT_VERSION` and `DEVICE_COMPACT_MESSAGE` is now supported by trunk, and it could be very helpful here, IMHO.

That's a good point, as I'm still not sure whether we should support migration from the sins of the past for ath79.

Hurricos commented:
> > `DEVICE_COMPACT_VERSION` and `DEVICE_COMPACT_MESSAGE` is now supported by trunk, and it could be very helpful here, IMHO.
> 
> That's a good point, as I'm still not sure whether we should support migration from the sins of the past for ath79.

Thanks a lot to @xabolcs for pointing me back to previous issues on this. I agree that supporting migration from the past is outside of the scope of this issue alone, but I've tested my procedure once by hand and confirmed it works. I had a flood last week which slowed things down (Merakis were shell-down so they floated and stayed dry believe it or not, but the power supplies I need to potentially replace).

Since a sysupgrade already requires manual fiddling (specifically for us to kick the beginning of the image forward by 128K), it makes sense to just include the MAC fix there too.

adrianschmutzler commented:
>        setenv bootcmd bootm 0xbf0a0000; saveenv;
>        tftpboot 0c00000 <filename-of-initramfs-kernel>.bin;
>        bootm 0c00000;

And I still wonder whether we could boot without modifying bootcmd at all when just having the firmware/kernel partition start at 0xc0000 / 0xbf0c0000 ?

Hurricos commented:
> > ```
> >    setenv bootcmd bootm 0xbf0a0000; saveenv;
> >    tftpboot 0c00000 <filename-of-initramfs-kernel>.bin;
> >    bootm 0c00000;
> > ```
> 
> And I still wonder whether we could boot without modifying bootcmd at all (theoretically starting from a fresh device) by just having the firmware/kernel partition start at 0xc0000 / 0xbf0c0000 ?

My tftpboot in this situation was simply copying somewhere into memory. It wasn't intentional. The bootcmd originally points to a second partition, I believe at 0xbfda0000? This is why 19.07 does this. The plan now is to cover up Meraki's incompatible decision.

xabolcs commented:
> And I still wonder whether we could boot without modifying bootcmd at all ...

Sure with `mtd-concat`: `part1` / `part2` (whichever uses the OEM firmware as primary partition) and `storage`.

Hurricos commented:
> > And I still wonder whether we could boot without modifying bootcmd at all ...
> 
> Sure with `mtd-concat`: `part1` / `part2` (whichever uses the OEM firmware as primary partition) and `storage`.

Let me get set to verify, but it sounds like that wound end us up in exactly the configuration 19.07 is in.

adrianschmutzler commented:
> > > And I still wonder whether we could boot without modifying bootcmd at all ...
> > 
> > 
> > Sure with `mtd-concat`: `part1` / `part2` (whichever uses the OEM firmware as primary partition) and `storage`.
> 
> Let me get set to verify, but it sounds like that wound end us up in exactly the configuration 19.07 is in.

It would be good to know, and maybe we should put it into the Wiki somewhere if somebody checks, so the next ones doesn't have to do it again.

Hurricos commented:
We do have the OEM's codebase after all: https://github.com/riptidewave93/meraki-linux/blob/linux-3.4-r23-20150601/arch/mips/ath79/mach-meraki-mr16.c

Here's how Meraki initializes the partitions in the OEM:

```c
static struct mtd_partition mr16_spiflash_parts[] = {
	{
		.name	= "Uboot & env",
		.offset = 0x00000000,
		.size	= 0x00080000,
	},
	{
		.name	= "board-config",
		.offset	= 0x00080000,
		.size	= 0x00020000,
	},
	{
		.name	= "panic",
		.offset	= 0x000A0000,
		.size	= 0x00020000,
	},
	{
		.name	= "storage",
		.offset	= 0x000C0000,
		.size	= 0x00200000,
	},
	{
		.name	= "caldata",
		.offset	= 0x00FE0000,
		.size	= 0x00020000,
	},
	{
		.name	= "part1",
		.offset	= 0x002C0000,
		.size	= 0x00680000,
	},
	{
		.name	= "part2",
		.offset	= 0x00940000,
		.size	= 0x00680000,
	},
	{
		.name	= "ALL",
		.offset	= 0x00000000,
		.size	= 0x01000000,
	},
};
```

... and while I don't have an unflashed Meraki on hand, I've got one that has only had its bootcmd changed (for  easy testing):

```sh
ar7100> printenv
bootargs=console=ttyS0,115200 root=31:02 rootfstype=jffs2 init=/sbin/init mtdparts=ar7100-nor0:256k(uboot),256k(env),6144k(rootfs),2048k(uImage),7680k(caldata) mem=64M
bootdelay=1
baudrate=115200
ipaddr=192.168.1.2
serverip=192.168.1.101
loadUboot=tftpboot 0x80010000 u-boot.bin;erase 0xbf000000 +0x40000;cp.b 0x80010000 0xbf000000 0x40000
loadLinux=tftpboot 0x80010000 vmlinux.gz.uImage;erase 0xbf680000 +0x200000;cp.b 0x80010000 0xbf680000 0x200000
loadFiles=tftpboot 0x80010000 ap96-jffs2;erase 0xbf080000 +0x600000;cp.b 0x80010000 0xbf080000 0x600000
loadAll=run loadUboot;run loadLinux;run loadFiles
factory_boot=tftpboot 0x3000000 vmlinux.gz.uImage;bootm 0x3000000
bootcmd=tftpboot 0c00000 openwrt-ath79-generic-meraki_mr16-initramfs-kernel.bin; bootm 0c00000
ethaddr=00:18:0a:35:a1:6c
stdin=serial      
stdout=serial
stderr=serial
ethact=eth0

Environment size: 845/65532 bytes
ar7100>
```

Specifically, check out this bit:

```
loadLinux=tftpboot 0x80010000 vmlinux.gz.uImage;erase 0xbf680000 +0x200000;cp.b 0x80010000 0xbf680000 0x200000
loadFiles=tftpboot 0x80010000 ap96-jffs2;erase 0xbf080000 +0x600000;cp.b 0x80010000 0xbf080000 0x600000
```

Good proof that Meraki is putting the files before the kernel, so since we are putting the kernel first (correct order) we should definitely be changing the `bootcmd`.

xabolcs commented:
And how do the erase adresses relate to partitions' start addesses?

I'm asking because I do not see!

Hurricos commented:
> loadLinux=tftpboot 0x80010000 vmlinux.gz.uImage;erase 0xbf680000 +0x200000;cp.b 0x80010000 0xbf680000 0x200000

Well, I see that the kernel (vmlinuz.gz.uImage) is being placed at 0xbf680000 and not 0xbf0c0000, which means the stock bootcmd was probably `bootm 0xbf680000`, which is why it does indeed need to be changed anyways, and we don't need to do any crazy maneuvering to build an image that requires one less environment variable to be set.

xabolcs commented:
> Well, I see that the kernel (vmlinuz.gz.uImage) is being placed at 0xbf680000 and not 0xbf0c0000,

I saw that too, and it doesn't make me any sense to flashing stuffs into the middle of partitions. That's why I asked about the address relation.

What commands does your u-boot support?
For Meraki MR12 it [looks like there is a `cpckgo.b` in the `bootcmd` variable](https://forum.archive.openwrt.org/viewtopic.php?id=48461) which has the relevant addresses:

```
PB93 (ar7241 - Virian) U-boot
DRAM: 
sri
ar7240_ddr_initial_config(133): virian ddr1 init
#### TAP VALUE 1 = 0x2, 2 = 0x2 [0x1d0cc0: 0x21f7]
64 MB
Top of RAM usable for U-Boot at: 84000000
Reserving 272k for U-Boot at: 83fb8000
Reserving 192k for malloc() at: 83f88000
Reserving 44 Bytes for Board Info at: 83f87fd4
Reserving 36 Bytes for Global Data at: 83f87fb0
Reserving 128k for boot params() at: 83f67fb0
Stack Pointer at: 83f67f98
Now running in RAM - U-Boot at: 83fb8000
id read 0x100000ff
flash size 16MB, sector count = 256
Flash: 16 MB
*** Warning - bad CRC, using default environment

In:    serial
Out:   serial
Err:   serial
Net:   ag7240_enet_initialize...
Fetching MAC Address from 0x83fe7ea0
Virian External MII mode MDC CFG Value ==> 6
: cfg1 0xf cfg2 0x7014
eth0 link down
ATHRSF1_PHY: PHY unit 0x0, address 0x4, ID 0xd04e,
ATHRSF1_PHY: Port 0, Neg Success
ATHRSF1_PHY: unit 0 port 0 phy addr 4
eth0 up
eth0
RESET is un-pushed
Hit any key to stop autoboot:  0
ar7240> bdinfo
boot_params = 0x83F67FB0
memstart    = 0x80000000
memsize     = 0x04000000
flashstart  = 0x9F000000
flashsize   = 0x01000000
flashoffset = 0x0003A060
ethaddr     = 00:AA:BB:CC:DD:EE
ip_addr     = 192.168.1.1
baudrate    = 115200 bps
ar7240> printenv
bootargs=console=ttyS0,115200 root=31:02 rootfstype=jffs2 init=/sbin/init mtdparts=ar7240-nor0:256k(u-boot),256k(u-boot-env),6144k(rootfs),1024k(uImage),256k(NVRAM),8448k(ART)
bootcmd=cpckgo.b 0x9f2c0000 0x9f940000 0x80060000
bootdelay=1
baudrate=115200
ethaddr=0x00:0xaa:0xbb:0xcc:0xdd:0xee
ipaddr=192.168.1.1
serverip=192.168.1.101
loadUboot=tftpboot 0x80010000 u-boot.bin;erase 0x9f000000 +0x40000;cp.b 0x80010000 0x9f000000 0x40000
loadLinux=tftpboot 0x80010000 vmlinux.gz.uImage;erase 0x9f680000 +0x200000;cp.b 0x80010000 0x9f680000 0x200000
loadFiles=tftpboot 0x80010000 ap96-jffs2;erase 0x9f080000 +0x600000;cp.b 0x80010000 0x9f080000 0x600000
loadAll=run loadUboot;run loadLinux;run loadFiles
factory_boot=tftpboot 0x3000000 vmlinux.gz.uImage;bootm 0x3000000
stdin=serial
stdout=serial
stderr=serial
ethact=eth0
```

xabolcs commented:
Do you have a full flash backup of MR16? Or at least ask Cisco for it or for the `ap96-jffs2` and `vmlinux.gz.uImage` TFTP files?

xabolcs commented:
Thank you!

Those help a lot!

I hope `cpckgo.b` has [the same logic as `meraki_late_init`](https://github.com/riptidewave93/meraki-linux/blob/linux-3.4-r23-20150601/arch/mips/ath79/mach-meraki.c#L202-L205) in `mach-meraki.c`!

How does the original OEM firmware upgrade work? Does the Meraki have some GUI to update the firmware?
I'm asking because `loadLinux` (`"vmlinux.gz.uImage"`) and `loadFiles` (`"ap96-jffs2"`) look weird to me:
```
[ 1.132] 0x00000000-0x00080000 : "Uboot & env" 2x 256k
[ 1.174] 0x00080000-0x000a0000 : "board config" 128k
         0xbf080000-0x00680000 : "ap96-jffs2" 0x600000 6144k
[ 1.216] 0x000a0000-0x000c0000 : "panic" 128k
[ 1.258] 0x000c0000-0x002c0000 : "storage" 2048k
[ 1.329] 0x002c0000-0x007e0000 : "part1" 6656k = 6144k + 512k
         0xbf680000-0x00880000 : "vmlinux.gz.uImage" 0x200000 2048k
[ 1.371] 0x00940000-0x00e60000 : "part2" 6656k = 6144k + 512k
[ 1.300] 0x00fe0000-0x01000000 : "caldata" 128k
[ 1.413] 0x00000000-0x01000000 : "ALL"
```

I don't have a device. I'm just reading some code and Wiki page, but here are my toughts about Adrian's "fresh device" question:
> And I still wonder whether we could boot without modifying bootcmd at all ...

Maintaining the possibility to rollback to OEM firmware without extra checks, OpenWrt have to occupy the `part1` partition only!
Rolling back to OEM firmware should be as easy as corrupting the `part1` partition and restarting the device: the bootloader starts the OEM kernel from `part2` which uses the `storage` partition.

If one could verify that the OEM firmware is able to repair a corrupted `storage` partition, then the previous 6656 kB `part1` partition could be `mtd-concat`-ed with `storage` to gain another 2048 kB flash space.

If OpenWrt occupies both `part1` and `part2` then it's allowed to occupy the `storage` partition too, as the OEM firmware is overwritten completely.

riptidewave93 commented:
So for the fun partition fun/mess, I can explain that. The MR12 was the first board I ever ported to OpenWRT back in 2014. At the time, I was not aware that the device mac came from the board config partition (called board-config on UBI meraki devices), since u-boot was not relying on it for the eth mac. This is what caused all MR12's to use the same mac, so the workaround was to add the "mac" partition so users can put it back to the correct mac since there was no way to revert post flashing. This quirk only exists on the MR12/MR16 meraki boards, since all other Meraki devices ported use the correct board-config partition for mac address info.

Surprised no one pinged me ahead of time to ask for this info. As for moving forward with this port, we can either keep the same quirk in place that exists in ar71xx currently (aka the mac partition), or we can try to move to the original flash layout but this will then cause issues for people already on the ar71xx image.

Hurricos committed:
ath79: add support for Meraki MR16

Port device support for Meraki MR16 from the ar71xx target to ath79.

Specifications:

  * AR7161 CPU, 16 MiB Flash, 64 MiB RAM
  * One PoE-capable Gigabit Ethernet Port
  * AR9220 / AR9223 (2x2 11an / 11n) WLAN

Installation:

  * Requires TFTP server at 192.168.1.101, w/ initramfs & sysupgrade .bins
  * Open shell case and connect a USB to TTL cable to upper serial headers
  * Power on the router; connect to U-boot over 115200-baud connection
  * Interrupt U-boot process to boot Openwrt by running:
       setenv bootcmd bootm 0xbf0a0000; saveenv;
       tftpboot 0c00000 <filename-of-initramfs-kernel>.bin;
       bootm 0c00000;
  * Copy sysupgrade image to /tmp on MR16
  * sysupgrade /tmp/<filename-of-sysupgrade>.bin

Notes:

  - There are two separate ARTs in the partition (offset 0x1000/0x5000 and
    0x11000/0x15000) in the OEM device. I suspect this is an OEM artifact;
    possibly used to configure the radios for different regions,
    circumstances or RF frontends. Since the ar71xx target uses the
    second offsets, use that second set (0x11000 and 0x15000) for the ART.

  - kmod-owl-loader is still required to load the ART partition into the
    driver.

  - The manner of storing MAC addresses is updated from ar71xx; it is
    at 0x66 of the 'config' partition, where it was discovered that the
    OEM firmware stores it. This is set as read-only. If you are
    migrating from ar71xx and used the method mentioned above to
    upgrade, use kmod-mtd-rw or UCI to add the MAC back in. One more
    method for doing this is described below.

  - Migrating directly from ar71xx has not been thoroughly tested, but
    one method has been used a couple of times with good success,
    migrating 18.06.2 to a full image produced as of this commit. Please
    note that these instructions are only for experienced users, and/or
    those still able to open their device up to flash it via the serial
    headers should anything go wrong.

    1) Install kmod-mtd-rw and uboot-envtools
    2) Run `insmod mtd-rw.ko i_want_a_brick=1`
    3) Modify /etc/fw_env.config to point to the u-boot-env partition.
       The file /etc/fw_env.config should contain:

       # MTD device   env offset  env size    sector size
       /dev/mtd1      0x00000     0x10000     0x10000

       See https://openwrt.org/docs/techref/bootloader/uboot.config
       for more details.

    4) Run `fw_printenv` to verify everything is correct, as per the
       link above.
    5) Run `fw_setenv bootcmd bootm 0xbf0a0000` to set a new boot address.
    6) Manually modify /lib/upgrade/common.sh's get_image function:
       Change ...

       cat "$from" 2>/dev/null | $cmd

       ... into ...

       (
         dd if=/dev/zero bs=1 count=$((0x66)) ; # Pad the first 102 bytes
         echo -ne '\x00\x18\x0a\x12\x34\x56'  ; # Add in MAC address
         dd if=/dev/zero bs=1 count=$((0x20000-0x66-0x6)) ; # Pad the rest
         cat "$from" 2>/dev/null | $cmd
       )

       ... which, during the upgrade process, will pad the image by
       128K of zeroes-plus-MAC-address, in order for the ar71xx's
       firmware partition -- which starts at 0xbf080000 -- to be
       instead aligned with the ath79 firmware partition, which
       starts 128K later at 0xbf0a0000.

    7) Copy the sysupgrade image into /tmp, as above
    8) Run `sysupgrade -F /tmp/<sysupgrade>.bin`, then wait

    Again, this may BRICK YOUR DEVICE, so make *sure* to have your
    serial cable handy.

Addenda:

  - The MR12 should be able to be migrated in a nearly identical manner as
    it shares much of its hardware with the MR16.

  - Thank-you Chris B for copious help with this port.

Signed-off-by: Martin Kennedy <hurricos@gmail.com>

Hurricos commented:
OK. Last push. I've added detailed instructions for manually articulating the upgrade from ar71xx to ath79, and I believe resolved all of the above issues, but correct me if I'm wrong.

Hurricos commented:
I know I said last push, but I'd like to make sure I clarify the u-boot-envtools step in the commit so it's easy for others to reproduce.

rotanid commented on target/linux/ath79/base-files/etc/uci-defaults/05_fix-compat-version:
```diff
@@ -0,0 +1,14 @@
+		;;
+esac
+
+exit 0
```
as GitHub also marks here, this file has no newline at its end  (unlike e.g. the dts file in this PR)

Hurricos replied:
Fixed and pushed.

riptidewave93 commented:
+1 From me

adschm commented:
Thanks for your work. I think the recent version is a good solution, dropping a few remnant of the past, but still keeping it manageable.

I will merge it later today.
